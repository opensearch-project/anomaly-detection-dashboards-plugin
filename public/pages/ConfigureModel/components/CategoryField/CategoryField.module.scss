.categoryFieldComboBox {
  // Ensure the combo box can handle multiple selections
  .euiComboBoxPill {
    max-width: 100%;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  .euiComboBox__inputWrap {
    max-height: 150px; // Adjust height to accommodate more selections
    overflow-y: auto;
  }
}
