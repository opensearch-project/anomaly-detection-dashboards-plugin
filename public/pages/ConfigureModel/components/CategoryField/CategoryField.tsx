/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  EuiFlexItem,
  EuiFlexGroup,
  EuiText,
  EuiLink,
  EuiIcon,
  EuiCompressedFormRow,
  EuiCompressedComboBox,
  EuiCompressedCheckbox,
  EuiTitle,
  EuiCallOut,
  EuiSpacer,
} from '@elastic/eui';
import { Field, FieldProps, FormikProps } from 'formik';
import { get, isEmpty } from 'lodash';
import { BASE_DOCS_LINK } from '../../../../utils/constants';
import React, { useState, useEffect } from 'react';
import ContentPanel from '../../../../components/ContentPanel/ContentPanel';
import {
  isInvalid,
  getError,
  validateCategoryField,
} from '../../../../utils/utils';
import { ModelConfigurationFormikValues } from '../../models/interfaces';

interface CategoryFieldProps {
  isEdit: boolean;
  isHCDetector: boolean;
  categoryFieldOptions: string[];
  setIsHCDetector(isHCDetector: boolean): void;
  isLoading: boolean;
  formikProps: FormikProps<ModelConfigurationFormikValues>;
}

export function CategoryField(props: CategoryFieldProps) {
  const [enabled, setEnabled] = useState<boolean>(
    get(props, 'formikProps.values.categoryFieldEnabled', false)
  );
  const noCategoryFields = isEmpty(props.categoryFieldOptions);
  const convertedOptions = props.categoryFieldOptions.map((option: string) => {
    return {
      label: option,
    };
  });

  useEffect(() => {
    // only update this if we're editing and the detector has finally come
    if (props.isEdit) {
      setEnabled(props.isHCDetector);
    }
  }, [props.isHCDetector]);

  return (
    <ContentPanel
      title={
        <EuiTitle size="s" id={'categoryFieldTitle'}>
          <h2>Categorical fields </h2>
        </EuiTitle>
      }
      subTitle={
        <EuiText
          className="content-panel-subTitle"
          style={{ lineHeight: 'normal' }}
        >
          Split a single time series into multiple time series based on
          categorical fields. You can select up to 2.{' '}
          <EuiLink href={`${BASE_DOCS_LINK}/ad`} target="_blank">
            Learn more
          </EuiLink>
        </EuiText>
      }
    >
      {noCategoryFields && !props.isLoading ? (
        <EuiCallOut
          data-test-subj="noCategoryFieldsCallout"
          title="There are no available category fields for the selected index"
          color="warning"
          iconType="alert"
          size="s"
        ></EuiCallOut>
      ) : null}
      {noCategoryFields ? <EuiSpacer size="m" /> : null}
      {props.isEdit ? (
        <EuiCallOut
          data-test-subj="categoryFieldReadOnlyCallout"
          title="You can't change the category fields after you create the detector."
          color="primary"
          iconType="iInCircle"
          size="s"
        ></EuiCallOut>
      ) : null}
      {props.isEdit ? <EuiSpacer size="m" /> : null}
      <Field
        name="categoryField"
        validate={enabled ? validateCategoryField : null}
      >
        {({ field, form }: FieldProps) => (
          <EuiFlexGroup direction="column">
            <EuiFlexItem>
              <EuiCompressedCheckbox
                id={'categoryFieldCheckbox'}
                label="Enable categorical fields"
                checked={enabled}
                disabled={noCategoryFields || props.isEdit}
                onChange={() => {
                  if (!enabled) {
                    props.setIsHCDetector(true);
                  }
                  if (enabled) {
                    props.setIsHCDetector(false);
                    form.setFieldValue('categoryField', []);
                  }
                  setEnabled(!enabled);
                }}
              />
            </EuiFlexItem>
            {enabled && !props.isEdit ? (
              <EuiFlexItem>
                <EuiCallOut
                  data-test-subj="cannotEditCategoryFieldCallout"
                  title="You can't change the category fields after you create the detector. Make sure that you only select the fields necessary for your use case."
                  color="warning"
                  iconType="alert"
                  size="s"
                ></EuiCallOut>
              </EuiFlexItem>
            ) : null}
            {enabled && !noCategoryFields ? (
              <EuiFlexItem>
                <EuiCompressedFormRow
                  label="Field"
                  isInvalid={isInvalid(field.name, form)}
                  error={getError(field.name, form)}
                  helpText={`You can only apply the categorical fields to the 'ip' and 'keyword' OpenSearch data types.`}
                >
                  <EuiCompressedComboBox
                    data-test-subj="categoryFieldComboBox"
                    id="categoryField"
                    placeholder="Select your categorical fields"
                    options={convertedOptions}
                    onBlur={() => {
                      form.setFieldTouched('categoryField', true);
                    }}
                    onChange={(options) => {
                      const selection = options.map((option) => option.label);
                      if (!isEmpty(selection)) {
                        if (selection.length <= 2) {
                          form.setFieldValue('categoryField', selection);
                        }
                      } else {
                        form.setFieldValue('categoryField', []);
                      }
                    }}
                    selectedOptions={
                      field.value
                        ? field.value.map((value: any) => {
                            return {
                              label: value,
                            };
                          })
                        : []
                    }
                    singleSelection={false}
                    isClearable={true}
                    isDisabled={props.isEdit}
                  />
                </EuiCompressedFormRow>
              </EuiFlexItem>
            ) : null}
          </EuiFlexGroup>
        )}
      </Field>
    </ContentPanel>
  );
}
