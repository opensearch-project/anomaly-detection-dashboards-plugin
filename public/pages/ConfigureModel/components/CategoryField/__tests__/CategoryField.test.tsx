/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React, { Fragment } from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Form, Formik } from 'formik';
import { CategoryField } from '../CategoryField';

describe('<CategoryField /> spec', () => {
  test('renders the component when disabled', () => {
    const { container, queryByText, queryByTestId } = render(
      <Fragment>
        <Formik
          initialValues={{
            categoryField: [],
          }}
          onSubmit={() => {}}
        >
          <Fragment>
            <Form>
              <CategoryField
                isEdit={false}
                isHCDetector={false}
                categoryFieldOptions={['option 1', 'option 2']}
                setIsHCDetector={(isHCDetector: boolean) => {
                  return;
                }}
                isLoading={false}
                formikProps={{
                  values: {
                    categoryFieldEnabled: false,
                  },
                }}
              />
            </Form>
          </Fragment>
        </Formik>
      </Fragment>
    );
    expect(container).toMatchSnapshot();
    expect(queryByTestId('noCategoryFieldsCallout')).toBeNull();
    expect(queryByTestId('categoryFieldComboBox')).toBeNull();
    expect(queryByText('Enable categorical fields')).not.toBeNull();
  });
  test('renders the component when enabled', () => {
    const { container, queryByText, queryByTestId, getByTestId, getByText } =
      render(
        <Fragment>
          <Formik
            initialValues={{
              categoryField: [],
            }}
            onSubmit={() => {}}
          >
            <Fragment>
              <Form>
                <CategoryField
                  isEdit={false}
                  isHCDetector={true}
                  categoryFieldOptions={['a', 'b']}
                  setIsHCDetector={(isHCDetector: boolean) => {
                    return;
                  }}
                  isLoading={false}
                  formikProps={{
                    values: {
                      categoryFieldEnabled: true,
                    },
                  }}
                />
              </Form>
            </Fragment>
          </Formik>
        </Fragment>
      );
    expect(container).toMatchSnapshot();
    expect(queryByTestId('noCategoryFieldsCallout')).toBeNull();
    expect(queryByTestId('categoryFieldComboBox')).not.toBeNull();
    expect(queryByText('Enable categorical fields')).not.toBeNull();
    fireEvent.click(getByTestId('comboBoxToggleListButton'));
    getByText('a');
    getByText('b');
  });
  test('shows callout when there are no available category fields', () => {
    const { queryByText, queryByTestId } = render(
      <Fragment>
        <Formik
          initialValues={{
            categoryField: [],
          }}
          onSubmit={() => {}}
        >
          <Fragment>
            <Form>
              <CategoryField
                isEdit={false}
                isHCDetector={true}
                categoryFieldOptions={[]}
                setIsHCDetector={(isHCDetector: boolean) => {
                  return;
                }}
                isLoading={false}
                formikProps={{
                  values: {
                    categoryFieldEnabled: true,
                  },
                }}
              />
            </Form>
          </Fragment>
        </Formik>
      </Fragment>
    );
    expect(queryByTestId('noCategoryFieldsCallout')).not.toBeNull();
    expect(queryByTestId('categoryFieldComboBox')).toBeNull();
    expect(queryByText('Enable categorical fields')).not.toBeNull();
  });
  test('hides callout if component is loading', () => {
    const { queryByText, queryByTestId } = render(
      <Fragment>
        <Formik
          initialValues={{
            categoryField: [],
          }}
          onSubmit={() => {}}
        >
          <Fragment>
            <Form>
              <CategoryField
                isEdit={false}
                isHCDetector={true}
                categoryFieldOptions={[]}
                setIsHCDetector={(isHCDetector: boolean) => {
                  return;
                }}
                isLoading={true}
                formikProps={{
                  values: {
                    categoryFieldEnabled: true,
                  },
                }}
              />
            </Form>
          </Fragment>
        </Formik>
      </Fragment>
    );
    expect(queryByTestId('noCategoryFieldsCallout')).toBeNull();
    expect(queryByText('Enable categorical fields')).not.toBeNull();
  });
  test(`limits selection to a maximum of 2 entities`, () => {
    const { getAllByRole, getByTestId, queryByText } = render(
      <Fragment>
        <Formik
          initialValues={{
            categoryField: [],
          }}
          onSubmit={() => {}}
        >
          <Fragment>
            <Form>
              <CategoryField
                isEdit={false}
                isHCDetector={true}
                categoryFieldOptions={['a', 'b', 'c']}
                setIsHCDetector={(isHCDetector: boolean) => {
                  return;
                }}
                isLoading={false}
                formikProps={{
                  values: {
                    categoryFieldEnabled: true,
                  },
                }}
              />
            </Form>
          </Fragment>
        </Formik>
      </Fragment>
    );
    // open combo box
    fireEvent.click(getByTestId('comboBoxToggleListButton'));
    expect(queryByText('a')).not.toBeNull();
    expect(queryByText('b')).not.toBeNull();
    expect(queryByText('c')).not.toBeNull();

    // select top 3 options (a,b,c)
    fireEvent.click(getAllByRole('option')[0]);
    fireEvent.click(getAllByRole('option')[0]);
    fireEvent.click(getAllByRole('option')[0]);

    // close combo box
    fireEvent.click(getByTestId('comboBoxToggleListButton'));

    // the last selection (c) is still not selected
    expect(queryByText('a')).not.toBeNull();
    expect(queryByText('b')).not.toBeNull();
    expect(queryByText('c')).toBeNull();
  });
  test(`fields are readonly if editing`, () => {
    const { getByTestId, queryByText } = render(
      <Fragment>
        <Formik
          initialValues={{
            categoryField: [],
          }}
          onSubmit={() => {}}
        >
          <Fragment>
            <Form>
              <CategoryField
                isEdit={true}
                isHCDetector={true}
                categoryFieldOptions={['a', 'b', 'c']}
                setIsHCDetector={(isHCDetector: boolean) => {
                  return;
                }}
                isLoading={false}
                formikProps={{
                  values: {
                    categoryFieldEnabled: true,
                  },
                }}
              />
            </Form>
          </Fragment>
        </Formik>
      </Fragment>
    );
    // try to open combo box. options should not show since the box is readonly
    fireEvent.click(getByTestId('comboBoxToggleListButton'));
    expect(queryByText('a')).toBeNull();
    expect(queryByText('b')).toBeNull();
    expect(queryByText('c')).toBeNull();
  });
  test('shows warning callout if creating & category field enabled', () => {
    const { queryByTestId } = render(
      <Fragment>
        <Formik
          initialValues={{
            categoryField: [],
          }}
          onSubmit={() => {}}
        >
          <Fragment>
            <Form>
              <CategoryField
                isEdit={false}
                isHCDetector={true}
                categoryFieldOptions={[]}
                setIsHCDetector={(isHCDetector: boolean) => {
                  return;
                }}
                isLoading={false}
                formikProps={{
                  values: {
                    categoryFieldEnabled: true,
                  },
                }}
              />
            </Form>
          </Fragment>
        </Formik>
      </Fragment>
    );
    expect(queryByTestId('cannotEditCategoryFieldCallout')).not.toBeNull();
  });
  test('hides warning callout if creating & category field disabled', () => {
    const { queryByTestId } = render(
      <Fragment>
        <Formik
          initialValues={{
            categoryField: [],
          }}
          onSubmit={() => {}}
        >
          <Fragment>
            <Form>
              <CategoryField
                isEdit={false}
                isHCDetector={true}
                categoryFieldOptions={[]}
                setIsHCDetector={(isHCDetector: boolean) => {
                  return;
                }}
                isLoading={false}
                formikProps={{
                  values: {
                    categoryFieldEnabled: false,
                  },
                }}
              />
            </Form>
          </Fragment>
        </Formik>
      </Fragment>
    );
    expect(queryByTestId('cannotEditCategoryFieldCallout')).toBeNull();
  });
  test('shows info callout and hides warning callout if editing', () => {
    const { queryByTestId } = render(
      <Fragment>
        <Formik
          initialValues={{
            categoryField: [],
          }}
          onSubmit={() => {}}
        >
          <Fragment>
            <Form>
              <CategoryField
                isEdit={true}
                isHCDetector={true}
                categoryFieldOptions={[]}
                setIsHCDetector={(isHCDetector: boolean) => {
                  return;
                }}
                isLoading={false}
                formikProps={{
                  values: {
                    categoryFieldEnabled: true,
                  },
                }}
              />
            </Form>
          </Fragment>
        </Formik>
      </Fragment>
    );
    expect(queryByTestId('categoryFieldReadOnlyCallout')).not.toBeNull();
    expect(queryByTestId('cannotEditCategoryFieldCallout')).toBeNull();
  });
});
