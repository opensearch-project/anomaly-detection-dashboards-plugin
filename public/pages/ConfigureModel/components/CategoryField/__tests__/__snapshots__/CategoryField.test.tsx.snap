// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryField /> spec renders the component when disabled 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
      style="padding: 20px;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="padding: 0px;"
      >
        <div
          class="euiFlexItem"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
            data-test-subj="contentPanelTitle"
          >
            <div
              class="euiFlexItem"
            >
              <h2
                class="euiTitle euiTitle--small"
                id="categoryFieldTitle"
              >
                Categorical fields 
              </h2>
            </div>
          </div>
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem content-panel-subTitle"
              style="line-height: normal; max-width: 75%;"
            >
              <div
                class="euiText euiText--medium content-panel-subTitle"
                style="line-height: normal;"
              >
                Split a single time series into multiple time series based on categorical fields. You can select up to 2.
                 
                <a
                  class="euiLink euiLink--primary"
                  href="https://opensearch.org/docs/monitoring-plugins/ad"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Learn more
                  <svg
                    aria-hidden="true"
                    aria-label="External link"
                    class="euiIcon euiIcon--small euiIcon-isLoading euiLink__externalIcon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                    />
                  </svg>
                  <span
                    class="euiScreenReaderOnly"
                  >
                    (opens in a new tab or window)
                  </span>
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            />
          </div>
        </div>
      </div>
      <div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
        />
        <div
          style="padding: 10px 0px;"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            >
              <div
                class="euiCheckbox"
              >
                <input
                  class="euiCheckbox__input"
                  id="categoryFieldCheckbox"
                  type="checkbox"
                />
                <div
                  class="euiCheckbox__square"
                />
                <label
                  class="euiCheckbox__label"
                  for="categoryFieldCheckbox"
                >
                  Enable categorical fields
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`<CategoryField /> spec renders the component when enabled 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
      style="padding: 20px;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="padding: 0px;"
      >
        <div
          class="euiFlexItem"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
            data-test-subj="contentPanelTitle"
          >
            <div
              class="euiFlexItem"
            >
              <h2
                class="euiTitle euiTitle--small"
                id="categoryFieldTitle"
              >
                Categorical fields 
              </h2>
            </div>
          </div>
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem content-panel-subTitle"
              style="line-height: normal; max-width: 75%;"
            >
              <div
                class="euiText euiText--medium content-panel-subTitle"
                style="line-height: normal;"
              >
                Split a single time series into multiple time series based on categorical fields. You can select up to 2.
                 
                <a
                  class="euiLink euiLink--primary"
                  href="https://opensearch.org/docs/monitoring-plugins/ad"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Learn more
                  <svg
                    aria-label="External link"
                    class="euiIcon euiIcon--small euiLink__externalIcon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M13 8.5a.5.5 0 1 1 1 0V12a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h3.5a.5.5 0 0 1 0 1H4a1 1 0 0 0-1 1v8a1 1 0 0 0 1 1h8a1 1 0 0 0 1-1V8.5Zm-5.12.339a.5.5 0 1 1-.706-.707L13.305 2H10.5a.5.5 0 1 1 0-1H14a1 1 0 0 1 1 1v3.5a.5.5 0 1 1-1 0V2.72L7.88 8.838Z"
                    />
                  </svg>
                  <span
                    class="euiScreenReaderOnly"
                  >
                    (opens in a new tab or window)
                  </span>
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            />
          </div>
        </div>
      </div>
      <div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
        />
        <div
          style="padding: 10px 0px;"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            >
              <div
                class="euiCheckbox"
              >
                <input
                  checked=""
                  class="euiCheckbox__input"
                  id="categoryFieldCheckbox"
                  type="checkbox"
                />
                <div
                  class="euiCheckbox__square"
                />
                <label
                  class="euiCheckbox__label"
                  for="categoryFieldCheckbox"
                >
                  Enable categorical fields
                </label>
              </div>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiCallOut euiCallOut--warning euiCallOut--small"
                data-test-subj="cannotEditCategoryFieldCallout"
              >
                <div
                  class="euiCallOutHeader"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiCallOutHeader__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                    />
                  </svg>
                  <span
                    class="euiCallOutHeader__title"
                  >
                    You can't change the category fields after you create the detector. Make sure that you only select the fields necessary for your use case.
                  </span>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFormRow"
                id="random_html_id-row"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    aria-invalid="false"
                    class="euiFormLabel euiFormRow__label"
                    for="random_html_id"
                  >
                    Field
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    aria-describedby="random_html_id-help-0"
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    class="euiComboBox"
                    data-test-subj="categoryFieldComboBox"
                    role="combobox"
                  >
                    <div
                      class="euiFormControlLayout"
                    >
                      <div
                        class="euiFormControlLayout__childrenWrapper"
                      >
                        <div
                          class="euiComboBox__inputWrap euiComboBox__inputWrap-isClearable"
                          data-test-subj="comboBoxInput"
                          tabindex="-1"
                        >
                          <p
                            class="euiComboBoxPlaceholder"
                          >
                            Select your categorical fields
                          </p>
                          <div
                            class="euiComboBox__input"
                            style="font-size: 14px; display: inline-block;"
                          >
                            <input
                              aria-controls=""
                              data-test-subj="comboBoxSearchInput"
                              id="random_html_id"
                              role="textbox"
                              style="box-sizing: content-box; width: 2px;"
                              value=""
                            />
                            <div
                              style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                            />
                          </div>
                        </div>
                        <div
                          class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                        >
                          <button
                            aria-label="Open list of options"
                            class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                            data-test-subj="comboBoxToggleListButton"
                            type="button"
                          >
                            <svg
                              aria-hidden="true"
                              class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                              focusable="false"
                              height="16"
                              role="img"
                              viewBox="0 0 16 16"
                              width="16"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                              />
                            </svg>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="euiFormHelpText euiFormRow__text"
                    id="random_html_id-help-0"
                  >
                    You can only apply the categorical fields to the 'ip' and 'keyword' OpenSearch data types.
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
